File: Python/Restore IP Addresses.py
"""
Given a string containing only digits, restore it by returning all possible valid IP address combinations.
For example:
Given "25525511135",
return ["255.255.11.135", "255.255.111.35"]. (Order does not matter)
"""

class Solution:
    # @param s, a string
    # @return a list of strings
    def restoreIpAddresses(self, s):
        M = len(s)
        result = []
        for i in range(1,min(4,M-2)):
            for j in range(i+1,min(i+4,M-1)):
                for k in range(j+1,min(j+4,M)):
                    s1 = s[:i]
                    s2 = s[i:j]
                    s3 = s[j:k]
                    s4 = s[k:]
                    if self.isValid(s1) and self.isValid(s2) and self.isValid(s3) and self.isValid(s4):
                        result.append(s1 + '.' + s2 + '.' + s3 +'.' +s4)
        return result
        
    def isValid(self,s):
        if len(s) > 3 or len(s) == 0 or (s[0] == '0' and len(s) > 1) or int(s) > 255:
            return False
        return True

class Solution:
    def restoreIpAddresses(self, s):
        result = []
        self.restoreIpAddressesRecur(result, s, "", 0)
        return result
        
    def restoreIpAddressesRecur(self, result, s, current, dots):
        # pruning to improve performance
        if (4 - dots) * 3 < len(s):
            return
        if dots == 3:
            if self.isValid(s):
                result.append(current + s)
        else:
            for i in range(3):
                if len(s) > i and self.isValid(s[:i + 1]):
                    self.restoreIpAddressesRecur(result, s[i + 1:], current + s[:i + 1] + '.', dots + 1)
        
    def isValid(self, s):
        if len(s) == 0 or (s[0] == "0" and s != "0"):
            return False
        return int(s) < 256

File: Java/Restore IP Addresses.java
/*
Given a string containing only digits, restore it by returning all possible valid IP address combinations.

For example:
Given "25525511135",

return ["255.255.11.135", "255.255.111.35"]. (Order does not matter)
*/

public class Solution {
    public List<String> restoreIpAddresses(String s) {
        List<String> result = new ArrayList<>();
        String ip = "";
        dfs(s, result, ip, 0, 0);
        return result;
    }
    
    private void dfs(String s, List<String> result, String ip, int start, int length) {
        if(start == s.length() && length == 4) {
            result.add(ip.substring(0, ip.length() - 1));
            return;
        }
        
        /* last one is more digit than 3*/
        if((s.length() - start) > 3 * (4 - length)) return;
        /* all the remaining digits cannot be mapped to an ip */
        if((s.length() - start) < (4 - length)) return;
        
        int num = 0;
        for(int i = start ; i < Math.min(start + 3, s.length()) ; i ++) {
            num = (s.charAt(i) - '0') + num * 10;
            
            if(num <= 255) {
                ip += s.charAt(i);
                dfs(s, result, ip + '.', i + 1, length + 1);
            }
            /* can not have 0 as first digit of a number */
            if(num == 0) break;
        }
    }
}